src/material_pool.cpp

```diff
@@ -12,218 +12,85 @@ namespace wr
-		else
-		{
-			if (!m_texture_pool)
-			{
-				m_texture_pool = normal.m_pool;
-			}
-			m_normal = normal;
-			m_material_data.m_material_flags.m_has_normal_texture = true;
-			
-		}
-	}
-
-	void Material::UseNormalTexture(bool use_normal)
-	{
-		m_material_data.m_material_flags.m_has_normal_texture = use_normal;
-	}
-
-	void Material::SetRoughness(TextureHandle roughness)
+		               float alpha_constant,
+					   float double_sided): Material(pool)
 	{
-		if (roughness.m_pool != m_texture_pool
-			&& m_texture_pool)
-		{
-			m_rougness = m_texture_pool->GetDefaultRoughness();
-			LOGW("Textures in a material need to belong to the same texture pool");
-			m_material_data.m_material_flags.m_has_roughness_texture = false;
-		}
-		else
-		{
-			if (!m_texture_pool)
-			{
-				m_texture_pool = roughness.m_pool;
-			}
-			m_rougness = roughness;
-			m_material_data.m_material_flags.m_has_roughness_texture = true;
-		}
-	}
 
-	void Material::UseRoughnessTexture(bool use_roughness)
-	{
-		m_material_data.m_material_flags.m_has_roughness_texture = use_roughness;
-	}
+		SetTexture(MaterialTextureType::ALBEDO, albedo);
+		SetTexture(MaterialTextureType::NORMAL, normal);
+		SetTexture(MaterialTextureType::ROUGHNESS, roughness);
+		SetTexture(MaterialTextureType::METALLIC, metallic);
+		SetTexture(MaterialTextureType::AO, ao);
 
-	void Material::SetMetallic(TextureHandle metallic)
-	{
-		if (metallic.m_pool != m_texture_pool
-			&& m_texture_pool)
-		{
-			m_metallic = m_texture_pool->GetDefaultMetalic();
-			LOGW("Textures in a material need to belong to the same texture pool");
-			m_material_data.m_material_flags.m_has_metallic_texture = false;
-		}
-		else
-		{
-			if (!m_texture_pool)
-			{
-				m_texture_pool = metallic.m_pool;
-			}
-			m_metallic = metallic;
-			m_material_data.m_material_flags.m_has_metallic_texture = true;
-		}
+		SetConstant(MaterialConstantType::USE_ALPHA_CONSTANT, alpha_constant);
+		SetConstant(MaterialConstantType::IS_DOUBLE_SIDED, double_sided);
 	}
 
-	void Material::UseMetallicTexture(bool use_metallic)
-	{
-		m_material_data.m_material_flags.m_has_metallic_texture = use_metallic;
+	TextureHandle Material::GetTexture(MaterialTextureType type) { 
+		return m_textures[size_t(type) % size_t(MaterialTextureType::COUNT)]; 
 	}
 
-	void Material::SetAmbientOcclusion(TextureHandle ao)
+	void Material::SetTexture(MaterialTextureType type, TextureHandle handle)
 	{
-		if (ao.m_pool != m_texture_pool
-			&& m_texture_pool)
+		if (handle.m_pool != m_texture_pool || !m_texture_pool)
 		{
-			m_ao = m_texture_pool->GetDefaultAO();
+			ClearTexture(type);
 			LOGW("Textures in a material need to belong to the same texture pool");
-			m_material_data.m_material_flags.m_has_ao_texture = false;
 		}
 		else
 		{
-			if (!m_texture_pool)
-			{
-				m_texture_pool = ao.m_pool;
-			}
-			m_ao = ao;
-			m_material_data.m_material_flags.m_has_ao_texture = true;
+			m_textures[size_t(type) % size_t(MaterialTextureType::COUNT)] = handle;
```

Comment:
not a fan of the C-style casting everywhere honestly

Reply:
Try to be more explicit in your casts as well, increases readability in general.